### PR TITLE
Support for basic density functions

### LINKED ISSUE #134: Add library of compare functions
As with the 'r' functions (e.g. `rbinom`, `rexp` etc) it may be useful to provide out-of-the-box 'd' versions which give the density in the comparison function. We could either do the same functions as we have random deviates for, or target some known epi models specifically.

We should also probably deal with log explicitly (either `log = TRUE/FALSE` or make `dpois` and `ldpois`)

### PR DESCRIPTION
**Pull request by richfitz**
Added support just for the functions that we need in sircovid for now

Fixes #134

### CODE REVIEW

File: tests/testthat/test-densities.R [hunk 1]
```diff
@@ -0,0 +1,57 @@
+context("densities")
+
+## This one is the easiest:
+test_that("dpois agrees", {
```
Comment (johnlees) [inline tests/testthat/test-densities.R:4]: I think we should add an edge case test to all of these:
n = 0; p/lambda = 0; p = 1 are the ones I can think of (and perhaps combinations of these)

`dpois(1, 0, log = TRUE) = -Inf` is the kind of thing that motivated `rexp`

File: inst/include/dust/densities.hpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+template <typename T>
+T lfactorial(T x) {
+  return std::lgamma(x + 1);
+}
+
+template <typename T>
+T lbeta(T x, T y) {
+  return lgamma(x) + lgamma(y) - lgamma(x + y);
+}
```
Comment (johnlees) [inline inst/include/dust/densities.hpp:28]: I can't remember if we're C++14 or C++17, but if the latter this (unlogged) is in <cmath> as std::beta

File: inst/include/dust/densities.hpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
+#ifndef DUST_DENSITIES_HPP
+#define DUST_DENSITIES_HPP
+
+#include <cmath>
+
+namespace dust {
+
+template <typename T>
+T maybe_log(T x, bool log) {
```
Comment (johnlees) [inline inst/include/dust/densities.hpp:9]: `inline` this I'd say?

Comment (johnlees) [reply inst/include/dust/densities.hpp:9]: Possibly inline all the functions here

File: inst/include/dust/densities.hpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    size * std::log(prob) + x * std::log(1 - prob);
+  return maybe_log(ret, log);
+}
+
+// A note on this parametrisation:
+//
+//   prob = alpha / (alpha + beta)
+//   rho = 1 / (alpha + beta + 1)
+//
+// Where alpha and beta have (0, Inf) support
```
Comment (johnlees) [inline inst/include/dust/densities.hpp:56]: I'll take your word for it 🙂

**Review by johnlees — CHANGES_REQUESTED** [review]

File: inst/include/dust/densities.hpp (hunk 2 above)
Comment (richfitz) [reply inst/include/dust/densities.hpp:28]: We really need the logged version though, and that's not coming in either standard

File: inst/include/dust/densities.hpp (hunk 3 above)
Comment (richfitz) [reply inst/include/dust/densities.hpp:9]: All template functions are inline

File: inst/include/dust/densities.hpp (hunk 4 above)
Comment (richfitz) [reply inst/include/dust/densities.hpp:56]: Haha, this is from @edknock

File: inst/include/dust/densities.hpp (hunk 3 above)
Comment (johnlees) [reply inst/include/dust/densities.hpp:9]: 🤯

Comment (johnlees) [reply inst/include/dust/densities.hpp:9]: I feel like I should have known that

Comment (richfitz) [reply inst/include/dust/densities.hpp:9]: Seems to depend on the exact definition of "inline", per SO

File: inst/include/dust/densities.hpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+template <typename T>
+T dnbinom(int x, int size, T mu, bool log) {
+  const T prob = size / (size + mu);
+  if (x == 0 && size == 0) {
+    return maybe_log(0, log);
+  }
+  const T ret = std::lgamma(static_cast<T>(x + size)) -
+    std::lgamma(static_cast<T>(size)) -
+    lfactorial(static_cast<T>(x)) +
```
[the remark this answers is not in the document]
Comment (richfitz) [reply inst/include/dust/densities.hpp]: I don't think it's that unreadable as maths :) But I have dropped the lfactorial as it's not getting us much

File: tests/testthat/test-densities.R (hunk 1 above)
Comment (richfitz) [reply tests/testthat/test-densities.R:4]: good shout - added the cases I could think of here, can add more now or later as we find them

**Review by johnlees — APPROVED** [review]
